pkg/kv/kvserver/gc/gc.go

```diff
@@ -390,140 +457,481 @@ func processReplicatedKeyRange(
+	return counterUpdate, nil
+}
+
+// maybeFlushPendingBatches flushes accumulated GC requests. If clear range is
+// used and current data is eligible for deletion then pending points batch and
+// clear range request are flushed.
+// If there's not enough point keys to justify clear range, then point point
+// batches may be flushed:
+//   - if there's a single batch and it didn't reach desired size, nothing is
+//     flushed
+//   - if there's more than single points batch accumulated, then all batches
+//     except for last are flushed
+func (b *gcKeyBatcher) maybeFlushPendingBatches(
+	ctx context.Context,
+) (counterUpdate gcInfoUpdate, err error) {
+	// Find where in first points batch is start key and flush "prefix".
+	if b.clearRangeEnabled && b.clearRangeCounters.versionsAffected >= b.clearRangeMinKeys {
+		// Optionally flush parts of the first batch if it doesn't match
+		// end of the range.
+		batchLen := len(b.pointsBatches[0].batchGCKeys)
+		// Find a key that is equal or less than end of clear range key.
+		// Batch is sorted in the reverse order, timestamps are not relevant since
+		// we can't have the same key more than once within single points batch.
+		lastIdx := sort.Search(batchLen, func(i int) bool {
+			return b.pointsBatches[0].batchGCKeys[i].Key.Compare(b.clearRangeEndKey) < 0
+		})
+		// If lastIdx == 0 we can have a clear range request which goes back to
+		// previous key completely covering this range. If this is true then we don't
+		// need to handle this batch.
+		if lastIdx > 0 {
+			b.pointsBatches[0].batchGCKeys = b.pointsBatches[0].batchGCKeys[:lastIdx]
+			b.pointsBatches[0].gcBatchCounters = b.partialPointBatchCounters
+			upd, err := b.flushPointsBatch(ctx, &b.pointsBatches[0])
+			if err != nil {
+				return gcInfoUpdate{}, err
+			}
+			counterUpdate.add(upd)
+		}
+		b.pointsBatches = make([]pointsBatch, 1)
+
+		// Flush clear range.
+		// To reduce locking contention between keys, if we have a range that only
+		// covers multiple versions of a single key we could avoid latching range
+		// and fall back to poing GC behaviour where gc threshold guards against
+		// key interference. We do it by resetting end key and thus signalling
+		// gc request handler to skip locking and use StartKey.Next() as the end
+		// of the range.
+		endRange := b.clearRangeEndKey
+		if b.clearRangeCounters.keysAffected == 1 {
+			endRange = nil
 		}
+		if err := b.gcer.GC(ctx, nil, nil, nil, &roachpb.GCRequest_GCClearSubRangeKey{
+			StartKey:          b.clearRangeStartKey.Key,
+			StartKeyTimestamp: b.clearRangeStartKey.Timestamp,
+			EndKey:            endRange,
+		}); err != nil {
+			if errors.Is(err, ctx.Err()) {
+				return gcInfoUpdate{}, err
+			}
+			// Even though we are batching the GC process, it's
+			// safe to continue because we bumped the GC
+			// thresholds. We may leave some inconsistent history
+			// behind, but nobody can read it.
+			log.Warningf(ctx, "failed to GC keys with clear range: %v", err)
```

Comment:
We will fail to delete range key in turn, it won't expose any old data.

Reply:
But only because we check it during request evaluation, right? I think we discussed removing the server-side check if it turned out to be too expensive, did we ever benchmark that?

Reply:
I don't think we benchmarked how expensive search vs deletion is. When running GC threshold evaluations, local time to scan and to delete is in the same ballpark so eliminating one for point keys might be beneficial.

I think the solution could be not to remove range tombstones if some key deletions failed, or tracking oldest non deleted version and only remove range tombstones below that.

Reply:
Yeah, I think we discussed not GCing range keys if point key GC failed.

Reply:
To be clear, we can leave this for later and write up an issue or TODO for it. But it would be good to verify that the scan isn't too expensive, since we're holding a write latch while running it.